sprint3

llvbh opened the pull request.

llvbh committed:
sprint3

alesyaisachenko commented on src/main/java/courier/Courier.java:
```diff
@@ -0,0 +1,21 @@
+package courier;
```
лучше разложить по пакетам классы из пакета courier. Классы для сериализации можно положить в пакет pojo, классы с методами для отправки запросов можно положить в пакет api.

alesyaisachenko commented on src/main/java/courier/CourierClient.java:
```diff
@@ -0,0 +1,63 @@
+    private final String LOGIN = ROOT + "/login";
+    private final String COURIER = ROOT + "/{courierId}";
+
+    public boolean create(Courier courier){
```
create(Courier courier) - лучше  createCourier(Courier courier), это упрощает чтение кода теста, в этом случае сразу из кода ясно какой именно объект создается.

Коммент относится ко всему коду (для метода delete(int courierId), и для методов из класса OrderClient).

alesyaisachenko commented on src/main/java/courier/CourierClient.java:
```diff
@@ -0,0 +1,63 @@
+            .path("ok");
+    }
+
+    public String createNoParams(Courier courier){
```
лучше создать один метод по отправке запроса и все. Проверять ответ уже в самом тесте и не выделять для него отдельный метод, то есть не стоило выделать отдельны методы createNoParams(Courier courier)), createDublicatedLogin(Courier courier), ...

Такой подход применить ко всему коду.

Это нужно обязательно исправить.

alesyaisachenko commented on src/main/java/courier/CourierClient.java:
```diff
@@ -0,0 +1,63 @@
+            .then()
+            .log().all()
+            .assertThat()
+            .statusCode(201)
```
статус коды можно было взять из библиотеки:
import  static  org.apache.http.HttpStatus.*;

statusCode(SC_CREATED)

alesyaisachenko commented on src/test/java/CourierTest.java:
```diff
@@ -0,0 +1,46 @@
+    }
+    @Test
+    @DisplayName("Check courier add method With No Params")
+    public void addCourierWithNoParams() {
```
с точки зрения тест-дизайна не рекомендуется в одном тесте объединять негативные проверки (то есть не заполнить и логин и пароль). Нужно создать как минимум два теста. В первом не заполнить поле логин (остальные все поля заполнить), во втором тесте не заполнить пароль (остальные поля заполнить).

Это нужно обязательно исправить.

alesyaisachenko commented on src/test/java/CourierTest.java:
```diff
@@ -0,0 +1,46 @@
+        assertEquals("Недостаточно данных для создания учетной записи", created);
+    }
+    @Test
+    @DisplayName("Check courier add method With all Params")
```
этот тест нужно разбить на три теста (успешное добавление курьера, успешный логин, нельзя создать двух одинаковых курьеров).

Сейчас если тест в середине упадет -- то часть функционала не протестируется, это плохо.

Это нужно обязательно исправить.

alesyaisachenko commented on src/test/java/OrderTest.java:
```diff
@@ -0,0 +1,42 @@
+    }
+    @Test
+    @DisplayName("Check new order without color")
+    public void newOrderWithNoColor() {
```
тесты создания заказа нужно было реализовать через параметризацию.

Это нужно обязательно исправить.

Commit:
черновик

llvbh committed:
разбила на 2 пакета, добавила параметризацию, разделила тесты

llvbh committed:
разбила на 2 пакета, добавила параметризацию, разделила тесты

llvbh committed:
//Алеся, помогите пожалуйста, на уроках по параметризованным тестам, мы не создавали массив строк отдельно new String[],
    // а здесь помогло только добавление new String[] перед аргументами. Где можно подробее про это почитать ?
    //на уроке было так, что сейчас приводит к ошибке missing argument:  public static Object[][] getTextData() {
    //        return new Object[][] { {"test ", "test"} }; }

alesyaisachenko commented on src/test/java/CourierCreateTest.java:
```diff
@@ -0,0 +1,48 @@
+    public void checkCreateCourier(){
+        Courier courier = new Courier("Ibyrai_science1", "qwerty12", "Altynsarin");
+        boolean createCourier = courierClient.createCourier(courier, 201, "ok");
+        assertTrue(createCourier);
```
созданного курьера нужно удалять после прохождения теста, лучше всего это делать в @after. В этом случае даже если тест упадет, удаление все равно будет производиться

alesyaisachenko commented on src/test/java/OrderTest.java:
```diff
@@ -0,0 +1,55 @@
+        this.color = color;
+    }
+    @Parameterized.Parameters
+    public static Object[][] getColor() {
```
public static Object[][] getColor(), Object[][] -- говорит о том, что тип возвращаемого значения этого метода является массив объектов Object (массив состоит из объектов). 
{new String[]{"BLACK"} -- первый элемент массива, его тип  Object
{new String[] {"GREY"}}-- второй элемент массива, его тип  Object и т.д.

Другую реализацию можно посмотреть в тутореале:
https://www.tutorialspoint.com/junit/junit_parameterized_test.htm

alesyaisachenko commented on src/test/java/OrderTest.java:
```diff
@@ -0,0 +1,55 @@
+                {new String[]{"BLACK"}},
+                {new String[] {"GREY"}},
+                {new String[]{"BLACK", "GREY"}}
+        };
```
нужно еще добавить вариант: можно совсем не указывать цвет;

alesyaisachenko commented on src/test/java/OrderTest.java:
```diff
@@ -0,0 +1,55 @@
+    }
+    @Test
+    @DisplayName("Get all orders")
+    public void getAllOrders(){
```
Параметризованные тесты лучше выносить в отдельный класс.
Сейчас у тебя не параметризованный тест запускаются несколько раз подряд -- это трата ресурсов.

Поэтому для параметризованного и не параметризованного теста лучше создавать разные классы. Это фишка JUnit4 , в JUnit5 уже можно в одном классе размещать все тесты и не параметризованные тесты будут запускаться один раз.
